FirebaseSegmentation/Sources/SEGDatabaseManager.m

```diff
@@ -0,0 +1,419 @@
+                @"Failed to insert config record. Wrong number of give parameters, current "
+                @"number is %ld, correct number is 4.",
+                (long)values.count);
+    return NO;
+  }
+  NSString *SQL = [NSString stringWithFormat:@"INSERT INTO %@ (%@, %@, %@, %@) values (?, ?, ?, ?)",
+                                             kMainTableName, kMainTableColumnApplicationIdentifier,
+                                             kMainTableColumnCustomInstallationIdentifier,
+                                             kMainTableColumnFirebaseInstallationIdentifier,
+                                             kMainTableColumnAssociationStatus];
+
+  sqlite3_stmt *statement = [self prepareSQL:[SQL UTF8String]];
+  if (!statement) {
+    return NO;
+  }
+
+  NSString *aString = values[0];
+  if (![self bindStringToStatement:statement index:1 string:aString]) {
+    return [self logErrorWithSQL:[SQL UTF8String] finalizeStatement:statement returnValue:NO];
+  }
+  aString = values[1];
+  if (![self bindStringToStatement:statement index:2 string:aString]) {
+    return [self logErrorWithSQL:[SQL UTF8String] finalizeStatement:statement returnValue:NO];
+  }
+  aString = values[2];
+  if (![self bindStringToStatement:statement index:3 string:aString]) {
+    return [self logErrorWithSQL:[SQL UTF8String] finalizeStatement:statement returnValue:NO];
+  }
+  aString = values[3];
+  if (![self bindStringToStatement:statement index:4 string:aString]) {
+    return [self logErrorWithSQL:[SQL UTF8String] finalizeStatement:statement returnValue:NO];
+  }
+  if (sqlite3_step(statement) != SQLITE_DONE) {
+    return [self logErrorWithSQL:[SQL UTF8String] finalizeStatement:statement returnValue:NO];
+  }
+  sqlite3_finalize(statement);
+  return YES;
+}
+
+/// TODO: (Check if required). Clear the record of given namespace and package name
+/// before updating the table.
+- (void)deleteRecordFromMainTableWithCustomInstanceIdentifier:
+    (nonnull NSString *)customInstanceIdentifier {
+}
+
+/// TODO: (Check if required). Remove all the records from a config content table.
+- (void)deleteAllRecordsFromTable {
+}
+
+#pragma mark - helper
+- (BOOL)executeQuery:(const char *)SQL withParams:(NSArray *)params {
+  SEG_MUST_NOT_BE_MAIN_THREAD();
+  sqlite3_stmt *statement = [self prepareSQL:SQL];
+  if (!statement) {
+    return NO;
+  }
+
+  [self bindStringsToStatement:statement stringArray:params];
+  if (sqlite3_step(statement) != SQLITE_DONE) {
+    return [self logErrorWithSQL:SQL finalizeStatement:statement returnValue:NO];
+  }
+  sqlite3_finalize(statement);
+  return YES;
+}
+
+/// Params only accept TEXT format string.
+- (BOOL)bindStringsToStatement:(sqlite3_stmt *)statement stringArray:(NSArray *)array {
+  int index = 1;
+  for (NSString *param in array) {
+    if (![self bindStringToStatement:statement index:index string:param]) {
+      return [self logErrorWithSQL:nil finalizeStatement:statement returnValue:NO];
```

Comment:
You can retrieve the original SQL string from `statement` with this method: https://www.sqlite.org/c3ref/expanded_sql.html